File: php/divergent-modifications/good/TransactionRepository.php
<?php

class TransactionRepository {
    private $db;

    public function __construct($db) {
        $this->db = $db;
    }

    public function saveTransaction(int $userId, float $amount, string $type): int {
        $stmt = $this->db->prepare("INSERT INTO transactions (user_id, amount, type, created_at) VALUES (?, ?, ?, NOW())");
        $stmt->execute([$userId, $amount, $type]);
        return $this->db->lastInsertId();
    }

    public function getUserBalance(int $userId): float {
        $stmt = $this->db->prepare("SELECT SUM(amount) as balance FROM transactions WHERE user_id = ?");
        $stmt->execute([$userId]);
        $result = $stmt->fetch();
        return $result['balance'] ?? 0;
    }

    public function getMonthlyTransactions(int $userId, int $month, int $year): array {
        $stmt = $this->db->prepare("
            SELECT type, SUM(amount) as total, COUNT(*) as count
            FROM transactions
            WHERE user_id = ? AND MONTH(created_at) = ? AND YEAR(created_at) = ?
            GROUP BY type
        ");
        $stmt->execute([$userId, $month, $year]);
        return $stmt->fetchAll();
    }

    public function getYearlyIncome(int $userId, int $year): float {
        $stmt = $this->db->prepare("
            SELECT SUM(amount) as total_income
            FROM transactions
            WHERE user_id = ? AND type = 'income' AND YEAR(created_at) = ?
        ");
        $stmt->execute([$userId, $year]);
        return $stmt->fetch()['total_income'] ?? 0;
    }

    public function getYearlyDeductions(int $userId, int $year): float {
        $stmt = $this->db->prepare("
            SELECT SUM(amount) as total_deductions
            FROM transactions
            WHERE user_id = ? AND type = 'deduction' AND YEAR(created_at) = ?
        ");
        $stmt->execute([$userId, $year]);
        return $stmt->fetch()['total_deductions'] ?? 0;
    }
}

File: python/divergent-modifications/good/transaction_repository.py
from datetime import datetime

class TransactionRepository:
    def __init__(self, db):
        self.db = db

    def save_transaction(self, user_id: int, amount: float, transaction_type: str) -> int:
        cursor = self.db.cursor()
        cursor.execute(
            "INSERT INTO transactions (user_id, amount, type, created_at) VALUES (?, ?, ?, ?)",
            (user_id, amount, transaction_type, datetime.now())
        )
        self.db.commit()
        return cursor.lastrowid

    def get_user_balance(self, user_id: int) -> float:
        cursor = self.db.cursor()
        cursor.execute("SELECT SUM(amount) as balance FROM transactions WHERE user_id = ?", (user_id,))
        result = cursor.fetchone()
        return result[0] if result and result[0] is not None else 0

    def get_monthly_transactions(self, user_id: int, month: int, year: int):
        cursor = self.db.cursor()
        cursor.execute("""
            SELECT type, SUM(amount) as total, COUNT(*) as count
            FROM transactions
            WHERE user_id = ? AND strftime('%m', created_at) = ? AND strftime('%Y', created_at) = ?
            GROUP BY type
        """, (user_id, f"{month:02d}", str(year)))
        return cursor.fetchall()

    def get_yearly_income(self, user_id: int, year: int) -> float:
        cursor = self.db.cursor()
        cursor.execute("""
            SELECT SUM(amount) as total_income
            FROM transactions
            WHERE user_id = ? AND type = 'income' AND strftime('%Y', created_at) = ?
        """, (user_id, str(year)))
        result = cursor.fetchone()
        return result[0] if result and result[0] is not None else 0

    def get_yearly_deductions(self, user_id: int, year: int) -> float:
        cursor = self.db.cursor()
        cursor.execute("""
            SELECT SUM(amount) as total_deductions
            FROM transactions
            WHERE user_id = ? AND type = 'deduction' AND strftime('%Y', created_at) = ?
        """, (user_id, str(year)))
        result = cursor.fetchone()
        return result[0] if result and result[0] is not None else 0
